### PR TITLE
Run mypy in pre-commit and fix errors it found

### PR DESCRIPTION
**Pull request by michaelosthege**
+ Fixed refactoring bugs in `evo_aspirate`/`evo_dispense` commands
+ Fixes type hints across the codebase
+ Added mypy to the pre-commit
+ Fixed warnings that got emitted in two tests
+ Version bumped to v1.7.1

### CODE REVIEW

File: robotools/evotools/commands.py [hunk 1]
```diff
@@ -236,20 +238,14 @@ def evo_aspirate(
...
     """
     # perform consistency checks
-    kwargs = dict(
+    (wells, labware_position, volume, liquid_class, tips,) = prepare_evo_aspirate_dispense_parameters(
         wells=wells,
         labware_position=labware_position,
         volume=volume,
         liquid_class=liquid_class,
         tips=tips,
+        max_volume=max_volume,
```
Comment (michaelosthege) [inline robotools/evotools/commands.py:247]: actual bug

File: robotools/evotools/commands.py [hunk 2]
```diff
@@ -310,20 +306,14 @@ def evo_dispense(
...
```
Comment (michaelosthege) [inline robotools/evotools/commands.py:315]: actual bug

File: robotools/evotools/worklist.py [hunk 3]
```diff
@@ -464,10 +468,11 @@ def aspirate_well(
             tube_id,
             rack_type,
             forced_rack_type,
-        ) = _prepare_aspirate_dispense_parameters(*args, max_volume=self.max_volume)
+            max_volume=self.max_volume,
```
Comment (michaelosthege) [inline robotools/evotools/worklist.py:471]: actual bug

File: robotools/evotools/worklist.py [hunk 4]
```diff
@@ -488,7 +493,8 @@ def evo_aspirate_well(
             stacklevel=2,
         )
         cmd = commands.evo_aspirate(
-            labware=labware,
+            n_rows=labware.n_rows,
+            n_columns=labware.n_columns,
```
Comment (michaelosthege) [inline robotools/evotools/worklist.py:497]: actual bug

File: robotools/evotools/worklist.py [hunk 5]
```diff
@@ -560,10 +565,11 @@ def dispense_well(
             tube_id,
             rack_type,
             forced_rack_type,
-        ) = _prepare_aspirate_dispense_parameters(*args, max_volume=self.max_volume)
+            max_volume=self.max_volume,
```
Comment (michaelosthege) [inline robotools/evotools/worklist.py:568]: actual bug

File: robotools/evotools/worklist.py [hunk 6]
```diff
@@ -584,7 +590,8 @@ def evo_dispense_well(
             stacklevel=2,
         )
         cmd = commands.evo_dispense(
-            labware=labware,
+            n_rows=labware.n_rows,
+            n_columns=labware.n_columns,
```
Comment (michaelosthege) [inline robotools/evotools/worklist.py:594]: actual bug

File: robotools/evotools/worklist.py [hunk 7]
```diff
@@ -855,7 +864,8 @@ def evo_aspirate(
         labware.remove(wells_calc, volumes_calc, label)
         self.comment(label)
         cmd = commands.evo_aspirate(
-            labware=labware,
+            n_rows=labware.n_rows,
+            n_columns=labware.n_columns,
```
Comment (michaelosthege) [inline robotools/evotools/worklist.py:868]: actual bug

File: robotools/evotools/worklist.py [hunk 8]
```diff
@@ -943,12 +953,14 @@ def evo_dispense(
         labware.remove(wells_calc, volumes_calc, label)
         self.comment(label)
         cmd = commands.evo_dispense(
-            labware=labware,
+            n_rows=labware.n_rows,
+            n_columns=labware.n_columns,
```
Comment (michaelosthege) [inline robotools/evotools/worklist.py:957]: actual bug

File: robotools/evotools/worklist.py [hunk 9]
```diff
@@ -943,12 +953,14 @@ def evo_dispense(
...
             wells=wells,
             labware_position=labware_position,
             volume=volumes,
             liquid_class=liquid_class,
             tips=tips,
+            max_volume=self.max_volume,
```
Comment (michaelosthege) [inline robotools/evotools/worklist.py:963]: actual bug

**Review by michaelosthege — COMMENTED** [review]
Mypy found actual bugs in 9 places(!)

**Review by rebeccahamel — APPROVED** [review]
I commented some general questions to expand my understanding of the code, but otherwise approve of this PR.

File: robotools/evotools/commands.py [hunk 10]
```diff
@@ -137,7 +138,8 @@ def prepare_evo_aspirate_dispense_parameters(
         raise ValueError("Missing required parameter: wells")
     if not isinstance(wells, (str, list, tuple, np.ndarray)):
         raise ValueError(f"Invalid wells: {wells}")
-    if not len(wells) == len(tips):
+    wells_list = list(np.atleast_1d(wells))
+    if not len(wells_list) == len(tips):
```
[the remark this answers is not in the document]
Comment (michaelosthege) [reply robotools/evotools/commands.py:142]: As per type annotation, the `wells` can also be a string `"A01"`, or a tuple `("A02", "C07")`.
The `np.atleast_1d` makes them into a `np.ndarray` but here we'd like to proceed with a `list[str]`.

However, it looks like we forgot to `.flatten("F")` the `np.ndarray`s of `wells` and `volumes` which could be 2D arrays provided by the user. I'll fix this.

File: robotools/evotools/commands.py [hunk 11]
```diff
@@ -96,11 +96,12 @@ def prepare_evo_aspirate_dispense_parameters(
...
     *,
     labware_position: Tuple[int, int],
-    volume: Union[float, List[float], int],
+    volume: Union[float, Sequence[float], int],
     liquid_class: str,
-    tips: Union[List[Tip], List[int]],
+    tips: Union[Sequence[Tip], Sequence[int]],
     max_volume: Optional[int] = None,
-) -> Tuple[list, tuple, float, str, list]:
+) -> Tuple[List[str], Tuple[int, int], List[float], str, List[Tip]]:
```
[the remark this answers is not in the document]
Comment (michaelosthege) [reply robotools/evotools/commands.py:103]: The return types should be informative to maximize downstream flexibility, and `List` is more informative than `Sequence`.

File: robotools/evotools/worklist.py [hunk 12]
```diff
@@ -723,22 +730,24 @@ def reagent_distribution(
...
             raise ValueError(f'"direction" must be either "left_to_right" or "right_to_left"')
-        direction = 0 if direction == "left_to_right" else 1
+        direction_i = 0 if direction == "left_to_right" else 1
 
         if exclude_wells is None:
-            exclude_wells = []
-        if len(exclude_wells) > 0:
+            exclude_list = []
+        else:
+            exclude_list = list(exclude_wells)
```
[the remark this answers is not in the document]
Comment (michaelosthege) [reply robotools/evotools/worklist.py:738]: The `reagent_distribution` method is the lower-level version that's not supposed to be used directly --- which we should indicate in the docstring, or by renaming the method 🙄

The `exclude_wells: Optional[Iterable[int]]` hint rules out scalars or 2D arrays so the `np.atleast_1d` is not needed here.

**Review by michaelosthege — COMMENTED** [review]
Thanks! Your questions made me look closer and I found two things that I'll take care of.

**Pull request merged**